librz/core/cil.c

```diff
@@ -424,3 +424,91 @@ RZ_IPI void rz_core_analysis_esil_default(RzCore *core) {
 	}
 	rz_core_seek(core, at, true);
 }
+
+/**********
+ * New Rizin IL Core (rzil) related things
+ *********/
+RZ_IPI void rz_core_analysis_rzil_init_mem(RzCore *core) {
+	RzILVM vm;
+	printf("[Init RzilVM memory\n");
+	if (core->analysis->rzil && core->analysis->rzil->vm) {
+		vm = core->analysis->rzil->vm;
+		rz_il_vm_add_mem(vm, vm->data_size);
+	}
+}
+
+RZ_IPI void core_rzil_init(RzCore *core) {
+	// real init;
+	// TODO : get args from config
+	//      currently it's bf specific
+	int addrsize = 64;
+	int datasize = 8;
+	ut64 start_addr = core->offset;
+
+	int romem = true;
+	int stats = true;
+	int nonull = true;
+
+	RzAnalysisRzil *rzil;
+	if (!(rzil = rz_analysis_rzil_new())) {
+		return;
+	}
+	// init
+	rz_il_vm_init(rzil->vm, start_addr, addrsize, datasize);
+	core->analysis->rzil = rzil;
+	// TODO : get registers info from rizin
+	rz_il_vm_add_reg(rzil->vm, "ptr", rzil->vm->addr_size);
+	rz_analysis_rzil_setup(core->analysis, rzil, romem, stats, nonull);
+}
+
+RZ_IPI void rz_core_analysis_rzil_init(RzCore *core) {
+	if (core->analysis->rzil) {
+		return;
+	}
+	core_rzil_init(core);
+}
+
+RZ_IPI void rz_core_analysis_rzil_reinit(RzCore *core) {
+	if (core->analysis->rzil) {
+		rz_analysis_rzil_cleanup(core->analysis, core->analysis->rzil);
+		core->analysis->rzil = NULL;
+	}
+
+	core_rzil_init(core);
+}
+
+// step a list of ct_opcode at a given address
+RZ_IPI void rz_core_rzil_step(RzCore *core) {
+	RzPVector *oplist;
+
+	if (!core->analysis || !core->analysis->rzil || !core->analysis->rzil->init_mem) {
+		return;
+	}
+
+	RzAnalysis *analysis = core->analysis;
+	RzAnalysisRzil *rzil = analysis->rzil;
+	RzILVM vm = rzil->vm;
+	RzAnalysisPlugin *cur = analysis->cur;
+	RzAnalysisOp op = { 0 };
+
+	if (!cur) {
+		// No analysis plugin
+		return;
+	}
+
+	ut64 addr = rz_il_bv_to_ut64(vm->pc);
+
+	// try load from vm
+	// fetch and parse if no opcode
+	ut8 code[32];
+	oplist = rz_il_vm_load_opcodes_at_pc(vm);
+	if (!oplist) {
+		// analysis current data to trigger rzil_set_op_code
+		(void)rz_io_read_at_mapped(core->io, addr, code, sizeof(code));
```

Comment:
I think it's better to handle errors here.

Reply:
Yeah it's another way to solve this. Currently I read from binary and analysis one opcode if there were none at pc_address. Would this approach cause problems ?

Reply:
Not for now, but maybe for sparse files, map boundaries, etc - it might.